prov/shm/src/smr_msg.c

```diff
@@ -94,29 +94,73 @@ static ssize_t smr_generic_recv(struct smr_srx_ctx *srx, const struct iovec *iov
 		uint64_t tag, uint64_t ignore, uint64_t flags,
 		struct smr_queue *recv_queue, struct smr_queue *unexp_queue)
 {
-	struct smr_rx_entry *entry;
-	ssize_t ret = -FI_EAGAIN;
-	fi_addr_t use_addr;
+	struct smr_match_attr match_attr;
+	struct smr_rx_entry *unexp, *owner_entry = NULL;
+	struct dlist_entry *dlist_entry;
+	bool buf_done = false;
+	int ret;
 
 	assert(iov_count <= SMR_IOV_LIMIT);
 	assert(!(flags & FI_MULTI_RECV) || iov_count == 1);
 
-	use_addr = srx->dir_recv ? addr : FI_ADDR_UNSPEC;
-	ret = smr_check_unexp_queue(srx, iov, desc, iov_count, use_addr,
-				    context, tag, ignore, flags, unexp_queue);
-	if (!ret)
-		return ret;
+	addr = srx->dir_recv ? addr : FI_ADDR_UNSPEC;
+	match_attr.id = addr;
+	match_attr.ignore = ignore;
+	match_attr.tag = tag;
 
 	ofi_spin_lock(&srx->lock);
+	if (flags & FI_MULTI_RECV) {
+		owner_entry = smr_get_recv_entry(srx, iov, desc, iov_count, addr,
+						 context, tag, ignore, flags);
+		owner_entry->peer_entry.size = ofi_total_iov_len(iov, iov_count);
+		if (!owner_entry) {
+			ret = -FI_ENOMEM;
+			goto out;
+		}
+	}
 
-	entry = smr_get_recv_entry(srx, iov, desc, iov_count, use_addr, context,
-				   tag, ignore, flags);
-	if (!entry) {
-		ret = -FI_ENOMEM;
-		goto out;
+	dlist_entry = dlist_remove_first_match(&unexp_queue->list,
+					       unexp_queue->match_func,
+					       &match_attr);
+
+	while (dlist_entry) {
+		unexp = container_of(dlist_entry, struct smr_rx_entry, peer_entry);
+		smr_init_rx_entry(unexp, iov, desc, iov_count, addr, context,
+				  tag, flags & (~FI_MULTI_RECV));
+		if (owner_entry) {
+			owner_entry->multi_recv_ref++;
+			unexp->peer_entry.owner_context = owner_entry;
+		}
+
+		if (owner_entry) {
+			if (smr_adjust_multi_recv(srx, &owner_entry->peer_entry,
+						  unexp->peer_entry.size))
+				buf_done = true;
+			iov = owner_entry->iov;
+		}
+
+		ofi_spin_unlock(&srx->lock);
+		ret = srx->peer_srx.peer_ops->start_msg(&unexp->peer_entry);
+		if (ret)
+			return ret;
+		else if (!owner_entry || buf_done)
+			return FI_SUCCESS;
```

Comment:
If done_buf is true, what happened to the original owner_entry?  Was that lost?

Reply:
The owner entry is freed in (free_entry) which is called in the start_msg processing/whenever the message is done so we don't care about freeing it here.